lib/libc/minimal/source/stdlib/malloc.c

```diff
@@ -0,0 +1,124 @@
+#include <stdlib.h>
+#include <zephyr.h>
+#include <init.h>
+#include <errno.h>
+#include <misc/mempool.h>
+#include <string.h>
+
+#if (CONFIG_MALLOC_ARENA_SIZE > 0)
+K_MUTEX_DEFINE(malloc_mutex);
+SYS_MEM_POOL_DEFINE(z_malloc_mem_pool, &malloc_mutex,  64,
+		    CONFIG_MALLOC_ARENA_SIZE, 1, 4, .data);
+
+void *malloc(size_t size)
+{
+	void *ret;
+
+	ret = sys_mem_pool_alloc(&z_malloc_mem_pool, size);
+	if (!ret) {
+		errno = ENOMEM;
+	}
+
+	return ret;
+}
+
+static int malloc_prepare(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+#ifdef CONFIG_USERSPACE
+	k_object_access_all_grant(&malloc_mutex);
+#endif
+	sys_mem_pool_init(&z_malloc_mem_pool);
+
+	return 0;
+}
+
+SYS_INIT(malloc_prepare, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#else /* No malloc arena */
+void *malloc(size_t size)
+{
+	ARG_UNUSED(size);
+
```

Comment:
Would be nice some debug info here informing that memory arena's size was not set.  This can avoid people digging into the code to see why memory allocation is failing.